node/bindnode/repr.go

```diff
@@ -0,0 +1,814 @@
+package bindnode
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	ipld "github.com/ipld/go-ipld-prime"
+	basicnode "github.com/ipld/go-ipld-prime/node/basic"
+	"github.com/ipld/go-ipld-prime/schema"
+)
+
+func reprNode(node ipld.Node) ipld.Node {
+	if node, ok := node.(schema.TypedNode); ok {
+		return node.Representation()
+	}
+	// ipld.Absent and ipld.Null are not typed.
+	// TODO: is this a problem? surely a typed struct's fields are always
+	// typed, even when absent or null.
+	return node
+}
+
+func reprStrategy(typ schema.Type) interface{} {
+	switch typ := typ.(type) {
+	case *schema.TypeStruct:
+		return typ.RepresentationStrategy()
+	case *schema.TypeUnion:
+		return typ.RepresentationStrategy()
+	}
+	return nil
+}
+
+type _prototypeRepr _prototype
+
+func (w *_prototypeRepr) NewBuilder() ipld.NodeBuilder {
+	return &_builderRepr{_assemblerRepr{
+		schemaType: w.schemaType,
+		val:        reflect.New(w.goType).Elem(),
+	}}
+}
+
+type _nodeRepr _node
+
+func (w *_nodeRepr) Kind() ipld.Kind {
+	switch stg := reprStrategy(w.schemaType).(type) {
+	case schema.StructRepresentation_Stringjoin:
+		return ipld.Kind_String
+	case schema.StructRepresentation_Map:
+		return ipld.Kind_Map
+	case schema.StructRepresentation_Tuple:
+		return ipld.Kind_List
+	case schema.UnionRepresentation_Keyed:
+		return ipld.Kind_Map
+	case schema.UnionRepresentation_Kinded:
+		haveIdx := int(w.val.FieldByName("Index").Int())
+		mtyp := w.schemaType.(*schema.TypeUnion).Members()[haveIdx]
+		return mtyp.TypeKind().ActsLike()
+	case schema.UnionRepresentation_Stringprefix:
+		return ipld.Kind_String
+	case nil:
+		return (*_node)(w).Kind()
+	default:
+		panic(fmt.Sprintf("TODO Kind: %T", stg))
+	}
+}
+
+func outboundMappedKey(stg schema.StructRepresentation_Map, key string) string {
```

Comment:
a comment explaining the need for inbound vs outbound here would be useful.

Reply:
These are pretty much just hacks, since the schema API gives the mappings one way but not the other. I'll add a TODO about working this out with Eric - I didn't add docs because I don't intend to maintain nor expose this code here.